File: 98. Validate Binary Search Tree/98. Validate Binary Search Tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_valid_bst(root)
    is_valid_bst_helper(root, -Float::INFINITY, Float::INFINITY)
end

def is_valid_bst_helper(node, min, max)
    return true if node.nil?
    
    return false if node.val <= min || node.val >= max
    
    is_valid_bst_helper(node.left, min, node.val) &&
    is_valid_bst_helper(node.right, node.val, max)
end

File: 98. Validate Binary Search Tree/98. Validate Binary Search Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
    return isValidBSTHelper(root, -Infinity, Infinity);
}

function isValidBSTHelper(node: TreeNode | null, min: number, max: number): boolean {
    if (node === null) return true;
    
    if (node.val <= min || node.val >= max) return false;
    
    return isValidBSTHelper(node.left, min, node.val) &&
           isValidBSTHelper(node.right, node.val, max);
}